### PR TITLE
Notification Filters

### LINKED ISSUE #4301: Notifications: Implement Filters
WordPress.com's Notifications tab has filters (Unread / Comments / Likes). Let's bring that feature to iOS as well.

### PR DESCRIPTION
**Pull request by jleandroperez**
#### Details:
In this PR we're implementing Notifications Filters, that allow our users to visualize only certain kind of notifications (All / Unread / Likes / Comments / Follows).

This has been implemented by means of the `tableHeaderView` property, which used to be (only) occupied by the `Ratings` view.

#### Scenario: Filters
1. Launch WPiOS
2. Tap over the Notifications tab
3. Verify that the new Segmented control (at the top) effectively applies the expected filters.

#### Scenario: Ratings
1. Comment out [this return statement](https://github.com/wordpress-mobile/WordPress-iOS/blob/issue/4301-notifications-filters/WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m#L308), so that the ratings view always shows up
2. Launch the app
3. Tap over the Notifications tab
4. The Ratings view should show on top (above the Filters).

--

Please, verify this feature on iPad / 5s / 6+

Needs Review: @aerych (Thank you Eric!!)
Closes #4301 

--

simulator screen shot oct 2 2015 4 14 42 pm

### CODE REVIEW

**Comment by sendhil** [comment]
Nice

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m [hunk 1]
```diff
@@ -47,6 +47,17 @@
 static NSTimeInterval NotificationsUndoTimeout          = 4;
 static NSString const *NotificationsNetworkStatusKey    = @"network_status";
 
+static CGFloat const RatingsViewHeight                  = 100.0f;
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m]: Super nitpicky but, maybe drop the `f` from the various `CGFloat`s and let the compiler worry about assigning the appropriate floating point flavor.

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m [hunk 2]
```diff
@@ -104,49 +120,18 @@ - (instancetype)initWithCoder:(NSCoder *)aDecoder
...
-    self.navigationItem.backBarButtonItem = backButton;
+    [self setupConstraints];
+    [self setupTableView];
+    [self setupTableHeaderView];
+    [self setupTableFooterView];
+    [self setupTableHandler];
+    [self setupRatingsView];
+    [self setupRefreshControl];
+    [self setupNavigationBar];
+    [self setupFiltersSegmentedControl];
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m]: Nice refactor :)

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m [hunk 3]
```diff
@@ -47,6 +47,17 @@
...
 
+static CGFloat const RatingsViewHeight                  = 100.0f;
+
+typedef NS_ENUM(NSUInteger, NotificationFilter)
+{
+    NotificationFilterNone      = 0,
+    NotificationFilterUnread    = 1,
+    NotificationFilterComment   = 2,
+    NotificationFilterFollow    = 3,
+    NotificationFilterLike      = 4
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m]: Any reason for the assignments? They would be the values of the enum even if not explicitly set, yah?

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m [hunk 4]
```diff
@@ -667,15 +785,28 @@ - (NSManagedObjectContext *)managedObjectContext
...
     return fetchRequest;
 }
 
+- (NSPredicate *)predicateForSelectedFilters
+{
+    NSDictionary *filtersMap = @{
+        @(NotificationFilterUnread)     : @" AND (read = NO)",
+        @(NotificationFilterComment)    : @" AND (type = 'comment')",
+        @(NotificationFilterFollow)     : @" AND (type = 'follow')",
+        @(NotificationFilterLike)       : @" AND (type = 'like')"
```
Comment (aerych) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m]: Might be nice to make `NoteTypeLike`, `NoteTypeFollow`, and `NoteTypeComment` in `Notification.m` external/public and reference the constants here.

**Comment by aerych** [comment]
Hi @jleandroperez :)  Cool addition!

I left a couple of nitpicky comments.  I did notice one thing while testing that seemed a bit off. 
If I have no unread notifiications, the "Unread" list will be empty when I look at it. (Fig 1.)  If I tap on a different tab then tap back to Notifications it shows a "No notifications yet" message. (Fig 2.) Then if I tap on a different notifications filter, the message remains and overlays the tableview. (Fig 3.)  Everything else looked great, including how it looks when multitasking on the iPad.

### Fig 1
simulator screen shot oct 5 2015 10 02 49 am

### Fig 2
simulator screen shot oct 5 2015 10 02 57 am

### Fig 3
simulator screen shot oct 5 2015 10 03 03 am

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m (hunk 1 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m]: Sorry about that one Eric, i really need to get used *not* to adding the `f`!!

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m (hunk 3 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m]: Just peace of mind! but it's time i dropped that. Updating, thank you!

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m (hunk 2 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m]: Thanks! i feel better now! :smile:

File: WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m (hunk 4 above)
Comment (jleandroperez) [inline WordPress/Classes/ViewRelated/Notifications/Controllers/NotificationsViewController.m]: Nice!! Updating right away, thanks Eric!

**Comment by jleandroperez** [comment]
Wooooow nice catch on the `Empty State` Eric!! (Thank you, missed that!). Addressing it shortly!

**Comment by jleandroperez** [comment]
@aerych **MANY** thanks for the review sir!!. Ready for another round!

**Comment by aerych** [comment]
Looks good @jleandroperez :)  I made one very nitpicky comment, but otherwise :shipit: 
Side note, I noticed when I look at an empty unread filter, the message layout is a bit wonky, particularly on wider screens.  I think that is more of an issue with the no results view and not something specific for this PR so I'll open an issue for us to tweak its layout where it makes sense.  
simulator screen shot oct 6 2015 8 46 40 am

**Comment by aerych** [comment]
Opened #4348

**Comment by jleandroperez** [comment]
Thanks a lot @aerych !!

**Pull request merged**